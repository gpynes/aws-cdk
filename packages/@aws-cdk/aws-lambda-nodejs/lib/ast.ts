import * as fs from 'fs';
import * as ts from 'typescript';

/**
 * An import statement
 */
interface Import {
  text: string;
  identifiers: string[];
}

/**
 * Creates a source file from a filename
 */
export function createSourceFile(filename: string): ts.SourceFile {
  return ts.createSourceFile(filename, fs.readFileSync(filename, 'utf8'), ts.ScriptTarget.Latest, true);
}

/**
 * Get top level import statements in a source file
 */
export function getImportStatements(sourceFile: ts.SourceFile): Import[] {
  const imports: Import[] = [];
  for (const statement of sourceFile.statements) {
    const importIdentifiers = getImportIdentifiers(statement);
    if (importIdentifiers) {
      imports.push({
        identifiers: importIdentifiers,
        text: statement.getText()
      });
    }
  }
  return imports;
}

function getImportIdentifiers(statement: ts.Statement): string[] | undefined {
  if (ts.isVariableStatement(statement)) { // JS
    for (const declaration of statement.declarationList.declarations) {
      if (declaration.initializer && ts.isCallExpression(declaration.initializer) && declaration.initializer.expression.getText() === 'require') {
        return collectIdentifiersUnder(declaration.name);
      }
    }
  }

  if (ts.isImportDeclaration(statement) || ts.isImportEqualsDeclaration(statement)) { // TS
    return collectIdentifiersUnder(statement);
  }

  return undefined;
}

/**
 * Collects all identifiers under a node
 */
export function collectIdentifiersUnder(node: ts.Node): string[] {
  let identifiers: string[] = [];
  if (ts.isIdentifier(node)) {
    identifiers.push(node.getText());
  }

  for (const child of node.getChildren()) {
    identifiers = identifiers.concat(collectIdentifiersUnder(child));
  }

  return identifiers;
}
