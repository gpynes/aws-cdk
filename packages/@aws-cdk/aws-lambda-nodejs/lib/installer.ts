import {spawnSync } from 'child_process';

interface InstallerOptions {
  readonly lockFile: string;
  readonly cacheGet: string[];
  readonly cacheEnv: string;
  readonly configEnv?: { [key: string]: string };
}

/**
 * A node module installer
 */
export class Installer {
  public static readonly NPM = new Installer('npm', {
    lockFile: 'package-lock.json',
    cacheGet: ['config', 'get', 'cache'],
    cacheEnv: 'npm_config_cache',
    configEnv: {
      'npm_config_loglevel': 'error', // Do not show WARN messages
      'npm_config_update-notifier': 'false', // Do not check for updates
    },
  });

  public static readonly YARN = new Installer('yarn', {
    lockFile: 'yarn.lock',
    cacheGet: ['cache', 'dir'],
    cacheEnv: 'YARN_CACHE_FOLDER',
  });

  public readonly command: string;

  public readonly options: InstallerOptions;

  constructor(command: string, options: InstallerOptions) {
    this.command = command;
    this.options = options;
  }

  public get cachePath(): string | undefined {
    try {
      const proc = spawnSync(this.command, this.options.cacheGet);

      if (proc.status !== 0) {
        return undefined;
      }

      return proc.stdout.toString().trim();
    } catch (err) {
      return undefined;
    }
  }
}
