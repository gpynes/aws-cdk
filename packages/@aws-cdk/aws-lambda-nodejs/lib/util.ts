import * as fs from 'fs';

// From https://github.com/errwischt/stacktrace-parser/blob/master/src/stack-trace-parser.js
const STACK_RE = /^\s*at (?:((?:\[object object\])?[^\\/]+(?: \[as \S+\])?) )?\(?(.*?):(\d+)(?::(\d+))?\)?\s*$/i;

/**
 * A parsed stack trace line
 */
export interface StackTrace {
  readonly file: string;
  readonly methodName?: string;
  readonly lineNumber: number;
  readonly column: number;
}

/**
 * Parses the stack trace of an error
 */
export function parseStackTrace(error?: Error): StackTrace[] {
  const err = error || new Error();

  if (!err.stack) {
    return [];
  }

  const lines = err.stack.split('\n');

  const stackTrace: StackTrace[] = [];

  for (const line of lines) {
    const results = STACK_RE.exec(line);
    if (results) {
      stackTrace.push({
        file: results[2],
        methodName: results[1],
        lineNumber: parseInt(results[3], 10),
        column: parseInt(results[4], 10),
      });
    }
  }

  return stackTrace;
}

/**
 * Returns the major version of node installation
 */
export function nodeMajorVersion(): number {
  return parseInt(process.versions.node.split('.')[0], 10);
}

/**
 * Finds closest package.json path
 */
export function findPkgPath(): string {
  let pkgPath;

  for (const path of module.paths) {
    pkgPath = path.replace(/node_modules$/, 'package.json');
    if (fs.existsSync(pkgPath)) {
      break;
    }
  }

  if (!pkgPath) {
    throw new Error('Cannot find a `package.json` file.');
  }

  return pkgPath;
}

/**
 * Counts items in a list
 */
export function countItems(list: string[]): { [item: string]: number } {
  const ret: { [identifier: string]: number } = {};
  for (const item of list) {
    if (ret[item]) {
      ret[item] += 1;
    } else {
      ret[item] = 1;
    }
  }

  return ret;
}
