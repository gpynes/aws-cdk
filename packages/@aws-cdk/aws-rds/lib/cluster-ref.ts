import ec2 = require('@aws-cdk/aws-ec2');
import cdk = require('@aws-cdk/cdk');
import { Endpoint } from './endpoint';

/**
 * Create a clustered database with a given number of instances.
 */
export interface IDatabaseCluster extends cdk.IConstruct, ec2.IConnectable {
  /**
   * Identifier of the cluster
   */
  readonly clusterIdentifier: string;

  /**
   * Identifiers of the replicas
   */
  readonly instanceIdentifiers: string[];

  /**
   * The endpoint to use for read/write operations
   */
  readonly clusterEndpoint: Endpoint;

  /**
   * Endpoint to use for load-balanced read-only operations.
   */
  readonly readerEndpoint: Endpoint;

  /**
   * Endpoints which address each individual replica.
   */
  readonly instanceEndpoints: Endpoint[];

  /**
   * The security group for this database cluster
   */
  readonly securityGroupId: string;

  /**
   * A dependable that can be depended upon to force cluster availability.
   */
  readonly available: cdk.IDependable;

  /**
   * Export a Database Cluster for importing in another stack
   */
  export(): DatabaseClusterImportProps;
}

/**
 * Properties that describe an existing cluster instance
 */
export interface DatabaseClusterImportProps {
  /**
   * The database port
   */
  port: string;

  /**
   * The security group for this database cluster
   */
  securityGroupId: string;

  /**
   * Identifier for the cluster
   */
  clusterIdentifier: string;

  /**
   * Identifier for the instances
   */
  instanceIdentifiers: string[];
  // Actual underlying type: DBInstanceId[], but we have to type it more loosely for Java's benefit.

  /**
   * Cluster endpoint address
   */
  clusterEndpointAddress: string;

  /**
   * Reader endpoint address
   */
  readerEndpointAddress: string;

  /**
   * Endpoint addresses of individual instances
   */
  instanceEndpointAddresses: string[];
}
