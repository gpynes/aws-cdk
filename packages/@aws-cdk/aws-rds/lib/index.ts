export * from './cluster';
export * from './cluster-ref';
export * from './instance';
export * from './props';
export * from './cluster-parameter-group';
export * from './endpoint';
export * from './instance';
export * from './option-group';
export * from './instance-parameter-group';
export * from './rotation-single-user';

// AWS::RDS CloudFormation Resources:
export * from './rds.generated';
