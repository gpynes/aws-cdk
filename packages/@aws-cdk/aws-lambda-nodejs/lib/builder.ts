import { spawnSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { findPkgPath } from './util';

/**
 * Builder options
 */
export interface BuilderOptions {
  /**
   * Entry file
   */
  readonly entry: string;

  /**
   * The output directory
   */
  readonly outDir: string;

  /**
   * Expose modules as UMD under this name
   */
  readonly global: string;

  /**
   * Minify
   */
  readonly minify?: boolean;

  /**
   * Include source maps
   */
  readonly sourceMaps?: boolean;

  /**
   * The cache directory
   */
  readonly cacheDir?: string;

  /**
   * The node version to use as target for Babel
   */
  readonly nodeVersion?: string;

  /**
   * External modules
   */
  readonly externals?: string[];

  /**
   * Modules that should be included in the `node_modules` directory
   */
  readonly includes?: string[];

  /**
   * Whether to install modules in Docker
   */
  readonly installInDocker?: boolean;
}

/**
 * Builder
 */
export class Builder {
  private readonly pkgPath: string;

  private readonly originalPkg: Buffer;

  private readonly originalPkgJson: { [key: string]: any };

  private readonly parcelBinPath: string;

  constructor(private readonly options: BuilderOptions) {
    // Original package.json
    this.pkgPath = findPkgPath();
    this.originalPkg = fs.readFileSync(this.pkgPath);
    this.originalPkgJson = JSON.parse(this.originalPkg.toString());

    // Find parcel
    let parcelPkgPath: string;
    try {
      parcelPkgPath = require.resolve('parcel-bundler/package.json'); // This will throw if `parcel-bundler` cannot be found
    } catch (err) {
      throw new Error('It looks like parcel-bundler is not installed. Please install v1.x of parcel-bundler with yarn or npm.');
    }
    const parcelDir = path.dirname(parcelPkgPath);
    const parcelPkg = JSON.parse(fs.readFileSync(parcelPkgPath, 'utf8'));

    if (!parcelPkg.version || !/^1\./.test(parcelPkg.version)) { // Peer dependency on parcel v1.x
      throw new Error(`This module has a peer dependency on parcel-bundler v1.x. Got v${parcelPkg.version}.`);
    }

    this.parcelBinPath = path.join(parcelDir, parcelPkg.bin.parcel);

    // Clean out dir
    if (fs.existsSync(this.options.outDir)) {
      fs.rmdirSync(this.options.outDir, { recursive: true });
    }
  }

  public build(): void {
    try {
      this.updatePkg();

      const args = [
        'build', this.options.entry,
        '--out-dir', this.options.outDir,
        '--out-file', 'index.js',
        '--global', this.options.global,
        '--target', 'node',
        '--bundle-node-modules',
        '--log-level', '2',
        !this.options.minify && '--no-minify',
        !this.options.sourceMaps && '--no-source-maps',
        ...this.options.cacheDir
          ? ['--cache-dir', this.options.cacheDir]
          : [],
      ].filter(Boolean) as string[];

      const parcel = spawnSync(this.parcelBinPath, args);

      if (parcel.error) {
        throw parcel.error;
      }

      if (parcel.status !== 0) {
        throw new Error(parcel.stdout.toString().trim());
      }

      if (this.options.includes) {
        this.installIncludes(this.options.includes);
      }
    } catch (err) {
      throw new Error(`Failed to build file at ${this.options.entry}: ${err}`);
    } finally { // Always restore package.json to original
      this.restorePkg();
    }
  }

  /**
   * Updates the package.json to configure Parcel
   */
  private updatePkg() {
    const updateData: { [key: string]: any } = {};
    if (this.options.nodeVersion) { // Update engines.node (Babel target)
      updateData.engines = { node: `>= ${this.options.nodeVersion}` };
    }

    if (this.options.externals) { // Add externals for parcel-plugin-externals
      updateData.externals = [...this.options.externals, ...this.options.includes || []];
    }

    if (Object.keys(updateData).length !== 0) {
      fs.writeFileSync(this.pkgPath, JSON.stringify({
        ...this.originalPkgJson,
        ...updateData,
      }, null, 2));
    }
  }

  /**
   * Install modules that should be included in the node_modules directory
   */
  private installIncludes(includes: string[]) {
    // Use original dependencies for versions
    const dependencies: { [dependency: string]: string } = {
      ...this.originalPkgJson.dependencies ?? {},
      ...this.originalPkgJson.devDependencies ?? {},
      ...this.originalPkgJson.peerDependencies ?? {},
    };

    // Retain only includes
    const filteredDependencies: { [dependency: string]: string } = {};
    for (const [d, v] of Object.entries(dependencies)) {
      if (includes.includes(d)) {
        filteredDependencies[d] = v;
      }
    }

    // Write dummy package.json
    fs.writeFileSync(path.join(this.options.outDir, 'package.json'), JSON.stringify({
      dependencies: filteredDependencies
    }, null, 2));

    let installer = Installer.NPM;

    // Check if we have lock files
    let lockFile: string | undefined;
    const yarnLock = path.join(path.dirname(this.pkgPath), LockFile.YARN);
    const pkgLock = path.join(path.dirname(this.pkgPath), LockFile.NPM);

    // If we have a `yarn.lock` file copy it and use `yarn` as installer
    // otherwise use `package-lock.json` and `npm`.
    if (fs.existsSync(yarnLock)) {
      installer = Installer.YARN;
      lockFile = yarnLock;
    } else if (fs.existsSync(pkgLock)) {
      lockFile = pkgLock;
    }

    if (lockFile) {
      fs.copyFileSync(lockFile, path.join(this.options.outDir, path.basename(lockFile)));
    }

    const command = this.options.installInDocker ? 'docker' : installer;
    const args = this.options.installInDocker
      ? ['run', '--rm', '-v', `${this.options.outDir}:/var/task`, `lambci/lambda:build-nodejs${this.options.nodeVersion || 12}.x`, 'npm', 'install']
      : ['install'];

    const install = spawnSync(command, args, {
        cwd: this.options.outDir
    });

    if (install.error) {
      throw install.error;
    }

    if (install.status !== 0) {
      throw new Error(install.stderr.toString().trim());
    }
  }

  private restorePkg() {
    fs.writeFileSync(this.pkgPath, this.originalPkg);
  }
}

enum Installer {
  NPM = 'npm',
  YARN = 'yarn',
}

enum LockFile {
  NPM = 'package-lock.json',
  YARN = 'yarn.lock',
}
