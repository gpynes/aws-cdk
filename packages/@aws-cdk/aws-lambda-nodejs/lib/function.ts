import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import { collectIdentifiersUnder, createSourceFile, getImportStatements } from './ast';
import { Builder } from './builder';
import { countItems, nodeMajorVersion, parseStackTrace } from './util';

/**
 * Options for a NodejsFunction
 */
export interface NodejsFunctionOptions extends lambda.FunctionOptions {
  /**
   * The runtime environment. Only runtimes of the Node.js family are
   * supported.
   *
   * @default - `NODEJS_12_X` if `process.versions.node` >= '12.0.0',
   * `NODEJS_10_X` otherwise.
   */
  readonly runtime?: lambda.Runtime;

  /**
   * Whether to minify files when bundling.
   *
   * @default false
   */
  readonly minify?: boolean;

  /**
   * Whether to include source maps when bundling.
   *
   * @default false
   */
  readonly sourceMaps?: boolean;

  /**
   * The build directory
   *
   * @default - `.build` in the entry file directory
   */
  readonly buildDir?: string;

  /**
   * The cache directory
   *
   * Parcel uses a filesystem cache for fast rebuilds.
   *
   * @default - `.cache` in the root directory
   */
  readonly cacheDir?: string;

  /**
   * A list of modules that should be considered as externals
   *
   * @default ['aws-sdk']
   */
  readonly externals?: string[];

  /**
   * A list of modules that should not be bundled but included in the
   * `node_modules` folder.
   *
   * @default - all modules are bundled
   */
  readonly includes?: string[]

  /**
   * Whether to install the modules in a docker environment
   *
   * @default false
   */
  readonly installInDocker?: boolean;
}

/**
 * Properties for a NodejsFunction
 */
export interface NodejsFunctionProps extends NodejsFunctionOptions {
  /**
   * Path to the entry file (JavaScript or TypeScript).
   *
   * @default - Derived from the name of the defining file and the construct's id.
   * If the `NodejsFunction` is defined in `stack.ts` with `my-handler` as id
   * (`new NodejsFunction(this, 'my-handler')`), the construct will look at `stack.my-handler.ts`
   * and `stack.my-handler.js`.
   */
  readonly entry?: string;

  /**
   * The name of the exported handler in the entry file.
   *
   * @default handler
   */
  readonly handler?: string;
}

/**
 * A Node.js Lambda function bundled using Parcel
 */
export class NodejsFunction extends lambda.Function {
  constructor(scope: cdk.Construct, id: string, props: NodejsFunctionProps = {}) {
    if (props.runtime && props.runtime.family !== lambda.RuntimeFamily.NODEJS) {
      throw new Error('Only `NODEJS` runtimes are supported.');
    }

    const entry = findEntry(id, props.entry);
    const handler = props.handler || 'handler';
    const buildDir = props.buildDir || path.join(path.dirname(entry), '.build');
    const handlerDir = path.join(buildDir, crypto.createHash('sha256').update(entry).digest('hex'));
    const defaultRunTime = nodeMajorVersion() >= 12
    ? lambda.Runtime.NODEJS_12_X
    : lambda.Runtime.NODEJS_10_X;
    const runtime = props.runtime || defaultRunTime;

    // Build with Parcel
    const builder = new Builder({
      entry,
      outDir: handlerDir,
      global: handler,
      minify: props.minify,
      sourceMaps: props.sourceMaps,
      cacheDir: props.cacheDir,
      nodeVersion: extractVersion(runtime),
      externals: props.externals ? Array.from(new Set(props.externals)) : ['aws-sdk'],
      includes: Array.from(new Set(props.includes)),
      installInDocker: props.installInDocker,
    });
    builder.build();

    super(scope, id, {
      ...props,
      runtime,
      code: lambda.Code.fromAsset(handlerDir),
      handler: `index.${handler}`,
    });
  }
}

/**
 * Properties for a NodejsCodeFunction
 */
export interface NodejsCodeFunctionProps extends NodejsFunctionOptions {
  /**
   * The handler's code.
   *
   * Must be declared as an anonymous function.
   *
   * @example
   * async (event) => {
   *   console.log(event);
   *   // do something with the event...
   *   return 'done';
   * };
   *
   * If the handler is imported from another file, use:
   * `async (event) => myImportedHandler(event);`
   */
  readonly handler: any;
}

/**
 * Node.js code function
 */
export class NodejsCodeFunction extends NodejsFunction {
  constructor(scope: cdk.Construct, id: string, props: NodejsCodeFunctionProps) {
    // Find defining file
    const definingFile = findDefiningFile('NodejsCodeFunction');
    const definingSourceFile = createSourceFile(definingFile);

    // Temporary file that will be used as entry
    const tmpFile = definingFile.replace(/.(js|ts)$/, '.tmp.$1');

    // Extract top level import/require statements from defining file
    const importStatements = getImportStatements(definingSourceFile);

    // Write import/require statements and handler's code to temporary file
    fs.writeFileSync(tmpFile, createHandlerCode(importStatements.map(s => s.text), props.handler));

    // Collect identifiers present in temporary files
    const tmpSourceFile = createSourceFile(tmpFile);
    const identifiers = collectIdentifiersUnder(tmpSourceFile);
    const identifiersCount = countItems(identifiers);

    // Filter out unused statement
    const usedImportStatements = importStatements.filter(statement => {
      for (const identifier of statement.identifiers) {
        if (identifiersCount[identifier] > 1 ) {
          return true;
        }
      }
      return false;
    });

    // Write a clean entry file that can be processed by the bundler
    fs.writeFileSync(tmpFile, createHandlerCode(usedImportStatements.map(s => s.text), props.handler));

    super(scope, id, {
      ...props,
      entry: tmpFile,
      handler: 'handler',
    });

    fs.unlinkSync(tmpFile);
  }
}

/**
 * Searches for an entry file. Preference order is the following:
 * 1. Given entry file
 * 2. A .ts file named as the defining file with id as suffix (defining-file.id.ts)
 * 3. A .js file name as the defining file with id as suffix (defining-file.id.js)
 */
function findEntry(id: string, entry?: string): string {
  if (entry) {
    if (!/\.(js|ts)$/.test(entry)) {
      throw new Error('Only JavaScript or TypeScript entry files are supported.');
    }
    if (!fs.existsSync(entry)) {
      throw new Error(`Cannot find entry file at ${entry}`);
    }
    return entry;
  }

  const definingFile = findDefiningFile();
  const extname = path.extname(definingFile);

  const tsHandlerFile = definingFile.replace(new RegExp(`${extname}$`), `.${id}.ts`);
  if (fs.existsSync(tsHandlerFile)) {
    return tsHandlerFile;
  }

  const jsHandlerFile = definingFile.replace(new RegExp(`${extname}$`), `.${id}.js`);
  if (fs.existsSync(jsHandlerFile)) {
    return jsHandlerFile;
  }

  throw new Error('Cannot find entry file.');
}

/**
 * Finds the name of the file where the `NodejsFunction` is defined
 */
function findDefiningFile(name: string = 'NodejsFunction'): string {
  const stackTrace = parseStackTrace();
  const functionIndex = stackTrace.findIndex(s => new RegExp(name).test(s.methodName || ''));

  if (functionIndex === -1 || !stackTrace[functionIndex + 1]) {
    throw new Error('Cannot find defining file.');
  }

  return stackTrace[functionIndex + 1].file;
}

/**
 * Extracts the version from the runtime
 */
function extractVersion(runtime: lambda.Runtime): string | undefined {
  const match = runtime.name.match(/nodejs(\d+)/);

  if (!match) {
    return undefined;
  }

  return match[1];
}

function createHandlerCode(imports: string[], handler: () => (event: any) => Promise<any>): string {
  return `
    ${imports.join('\n')}
    exports.handler=${handler};
  `;
}
