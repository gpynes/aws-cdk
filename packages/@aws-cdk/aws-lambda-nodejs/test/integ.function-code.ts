/* tslint:disable no-console */
import { App, Construct, Stack, StackProps } from '@aws-cdk/core';
import got from 'got';
import * as lambda from '../lib';
import { add, mult } from './math';

class TestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new lambda.NodejsCodeFunction(this, 'WithDeps', {
      handler: async (event: { a: number, b: number }) => {
        // Use local dependencies
        console.log(mult(event.a, event.b));
        console.log(add(event.a, event.b));

        // Use external dependency
        const body = await got('https://aws.amazon.com', {
          resolveBodyOnly: true
        });
        console.log(body);
      },
      includes: ['got'], // Do not bundle 'got'
      installInDocker: true, // Install it in a docker container
    });
  }
}

const app = new App();
new TestStack(app, 'cdk-integ-lambda-nodejs-from-code');
app.synth();
