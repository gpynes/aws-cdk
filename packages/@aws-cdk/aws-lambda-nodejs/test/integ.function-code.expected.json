{
  "Resources": {
    "WithDepsServiceRoleE7B2C978": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "WithDeps8C8F2EB0": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438bS3Bucket8AF32960"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438bS3VersionKey97DCB7C7"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438bS3VersionKey97DCB7C7"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "WithDepsServiceRoleE7B2C978",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "WithDepsServiceRoleE7B2C978"
      ]
    }
  },
  "Parameters": {
    "AssetParameters64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438bS3Bucket8AF32960": {
      "Type": "String",
      "Description": "S3 bucket for asset \"64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438b\""
    },
    "AssetParameters64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438bS3VersionKey97DCB7C7": {
      "Type": "String",
      "Description": "S3 key for asset version \"64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438b\""
    },
    "AssetParameters64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438bArtifactHash7DF8AC2B": {
      "Type": "String",
      "Description": "Artifact hash for asset \"64f143007e1d72a81f5b6ee51797e9c5d2d0b6a28af6483e55785daf49e7438b\""
    }
  }
}